Frame.cc

```diff
@@ -1370,66 +1379,105 @@ bool Frame::GetRegionSpectralData(std::vector<std::vector<double>>& stats_values
+    interrupt = false;
+    return interrupt;
 }
 
 bool Frame::Interrupt(int region_id, const RegionState& region_state) {
+    bool interrupt(true);
     if (!IsConnected()) {
         std::cerr << "[Region " << region_id << "] closing image, exit zprofile (statistics) before complete" << std::endl;
-        return true;
+        return interrupt;
     }
     if (!IsSameRegionState(region_id, region_state)) {
         std::cerr << "[Region " << region_id << "] region state changed, exit zprofile (statistics) before complete" << std::endl;
-        return true;
+        return interrupt;
     }
-    return false;
+    interrupt = false;
+    return interrupt;
 }
 
-bool Frame::Interrupt(int region_id, int profile_index, const RegionState& region_state, const std::vector<int>& requested_stats) {
+bool Frame::Interrupt(int region_id, int num_profiles, int profile_index, int profile_stokes, const RegionState& region_state,
+    const std::vector<int>& config_stats) {
+    bool interrupt(true);
     if (!IsConnected()) {
         std::cerr << "[Region " << region_id << "] closing image, exit zprofile (statistics) before complete" << std::endl;
-        return true;
+        return interrupt;
     }
     if (!IsSameRegionState(region_id, region_state)) {
         std::cerr << "[Region " << region_id << "] region state changed, exit zprofile (statistics) before complete" << std::endl;
-        return true;
+        return interrupt;
     }
-    if (!AreSameRegionSpectralRequests(region_id, profile_index, requested_stats)) {
+    if (!IsSameRegionSpectralConfig(region_id, num_profiles, profile_index, profile_stokes, config_stats)) {
         std::cerr << "[Region " << region_id << "] region requirement changed, exit zprofile (statistics) before complete" << std::endl;
-        return true;
+        return interrupt;
     }
-    return false;
+    interrupt = false;
+    return interrupt;
 }
 
 bool Frame::IsConnected() {
     return _connected;
 }
 
 bool Frame::IsSameRegionState(int region_id, const RegionState& region_state) {
-    return (_region_states.count(region_id) && _region_states[region_id] == region_state);
+    if (_regions.count(region_id)) {
+        return (_regions[region_id]->GetRegionState() == region_state);
+    }
+    return false;
 }
 
-bool Frame::AreSameRegionSpectralRequests(int region_id, int profile_index, const std::vector<int>& requested_stats) {
-    return (_region_requests.count(region_id) && _region_requests[region_id].IsAmong(profile_index, requested_stats));
+bool Frame::IsSameRegionSpectralConfig(
+    int region_id, int num_profiles, int profile_index, int profile_stokes, const std::vector<int>& config_stats) {
+    bool is_same(false);
+    if (_regions.count(region_id)) {
+        auto& region = _regions[region_id];
+        // check is number of profiles changed
+        size_t new_num_profiles(region->NumSpectralProfiles());
+        if ((new_num_profiles == 0) || (new_num_profiles != num_profiles)) {
```

Comment:
Sorry, I was not clear.  If calculating "z"/RMS and new_num_profiles=0, then always cancel (no profiles requested).  But if new requirements are "Qz"/Mean, "z"/RMS (num profiles=2 now) then the "z"/RMS profile is still needed (note its profile index changed).  If new requirements are "z"/[Mean, RMS] then "z"/RMS is still needed (note config stats changed).  If new requirements are "z"/Sum then "z"/RMS is no longer needed and should be cancelled.  So you have to check the region requirements for both the stokes and stats types to determine if the current profile is still needed.

Reply:
If we are calculating "z"/RMS and a new requirement "z"/[Mean, RMS] is set. Then the backend will cancel the old "z"/RMS calculation and start new "z"/[Mean, RMS] calculations. However, in this case, it is not necessary to cancel the old "z"/RMS and start a new exactly same calculation. In order to avoid this, the `_spectral_profiles` in `RegionProfiler` and its usage (for calculating stats data, filling the data in protobuf) may need to be changed. Should I finish this work in this PR, or start a another branch to do it?